dev/sg/sg_db.go

```diff
@@ -169,45 +169,76 @@ func dbAddUserAction(cmd *cli.Context) error {
 	if err != nil {
 		return err
 	}
+
 	db := database.NewDB(logger, conn)
+	err = db.WithTransact(ctx, func(tx database.DB) error {
+		username := cmd.String("username")
+		password := cmd.String("password")
+
+		// Create the user, generating an email based on the username.
+		email := fmt.Sprintf("%s@sourcegraph.com", username)
+		user, err := tx.Users().Create(ctx, database.NewUser{
+			Username:        username,
+			Email:           email,
+			EmailIsVerified: true,
+			Password:        password,
+		})
+		if err != nil {
+			return err
+		}
 
-	username := cmd.String("username")
-	password := cmd.String("password")
+		// Make the user site admin.
+		err = tx.Users().SetIsSiteAdmin(ctx, user.ID, true)
+		if err != nil {
+			return err
+		}
 
-	// Create the user, generating an email based on the username.
-	email := fmt.Sprintf("%s@sourcegraph.com", username)
-	user, err := db.Users().Create(ctx, database.NewUser{
-		Username:        username,
-		Email:           email,
-		EmailIsVerified: true,
-		Password:        password,
-	})
-	if err != nil {
-		return err
-	}
+		firstParam := 2
+		systemRoles, err := tx.Roles().List(ctx, database.RolesListOptions{
+			// Since we only have two system roles on every Sourcegraph instance, it's okay
+			// to hard-code a pagination parameter here.
+			PaginationArgs: &database.PaginationArgs{
+				First: &firstParam,
+			},
+			System: true,
+		})
+		if err != nil {
+			return err
+		}
 
-	// Make the user site admin.
-	err = db.Users().SetIsSiteAdmin(ctx, user.ID, true)
-	if err != nil {
-		return err
-	}
+		var roleIDs []int32
+		for _, role := range systemRoles {
+			roleIDs = append(roleIDs, role.ID)
+		}
 
-	// Report back the new user information.
-	std.Out.WriteSuccessf(
-		// the space after the last %s is so the user can select the password easily in the shell to copy it.
-		"User '%s%s%s' (%s%s%s) has been created and its password is '%s%s%s'.",
-		output.StyleOrange,
-		username,
-		output.StyleReset,
-		output.StyleOrange,
-		email,
-		output.StyleReset,
-		output.StyleOrange,
-		password,
-		output.StyleReset,
-	)
+		opts := database.BulkCreateForUserOpts{
+			UserID:  user.ID,
+			RoleIDs: roleIDs,
+		}
 
-	return nil
+		if _, err = tx.UserRoles().BulkCreateForUser(ctx, opts); err != nil {
+			return err
+		}
+
+		// Report back the new user information.
+		std.Out.WriteSuccessf(
+			// the space after the last %s is so the user can select the password easily in the shell to copy it.
+			"User '%s%s%s' (%s%s%s) has been created and its password is '%s%s%s'.",
+			output.StyleOrange,
+			username,
+			output.StyleReset,
+			output.StyleOrange,
+			email,
+			output.StyleReset,
+			output.StyleOrange,
+			password,
+			output.StyleReset,
+		)
```

Comment:
Golang string formatting supports indexed parameters, so you should be able to write this as

```suggestion
		std.Out.WriteSuccessf(
			// the space after the last %s is so the user can select the password easily in the shell to copy it.
			"User '%[1]s%[2]s%[3]s' (%[1]s%[2]s%[3]s) has been created and its password is '%[1]s%[2]s%[3]s'.",
			output.StyleOrange,
			username,
			output.StyleReset,
		)
```

Reply:
This doesn't work very well with our library. 

<img width="825" alt="CleanShot 2023-02-14 at 02 56 13@2x" src="https://user-images.githubusercontent.com/25608335/218618734-0d0a9656-0fc1-4d24-8ad2-a44d975475e8.png">

I kept getting weird results that appended one of the parameters to the end of the string like this.

```go
std.Out.WriteSuccessf(
        "User '%[2]s%[4]s%[3]s' (%[2]s%[5]s%[3]s) has been created and its password is '%[2]s%[6]s%[3]s'. \nThank you for using the tool.",
	output.StyleOrange,
	output.StyleReset,
	username,
	email,
	password,
)
```

Reply:
Ah okay I see, you would have to `Sprintf` the message before passing it to `WriteSuccessf` like:

```
std.Out.WriteSuccessf(
    fmt.Sprintf("User '%[1]s%[3]s%[2]s' (%[1]s%[4]s%[2]s) has been created and its password is '%[1]s%[5]s%[2]s'.\nThank you for using the tool.",
    output.StyleOrange,
    output.StyleSuccess,
    username,
    email,
    password,
))
```

(Notice I use `StyleSuccess` instead of `StyleReset` to change the colour back to Green like the start of the message).

The `WriteSuccessf` does weird things with the formatting. I'll leave the decision of whether you want to do it like this up to you